# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
path=os.getcwd()
path

'C:\\Users\\MONIKA'

In [3]:
path="D:\SPRINGBOARD_1\GuidedCapstone\Step 2\Step 2\data"
os.chdir(path)
os.listdir(path)

['step2_output.csv', 'step3_output.csv']

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [4]:
df=pd.read_csv(os.path.join(path,r'step3_output.csv') )
df.head(5)

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3         Boreal Mountain Resort  California         7700            500   
4                    Dodge Ridge  California         8200           1600   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          1     1       3  ...   
4      0        0.0          0          0     1       2  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              380.0      200.000000        150.000000       54.0   
4              862.0      174.873239        115.103943       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0      34.00000              152.0   
1            250.0          74.0      78.00000              104.0   
2            300.0          47.0      47.00000              107.0   
3            400.0          49.0      64.16681              150.0   
4            350.0          78.0      78.00000              140.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3           200.0         1  
4             0.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [5]:
dfo=df['state']
dfo

0          Alaska
1         Arizona
2      California
3      California
4      California
          ...    
171       Wyoming
172       Wyoming
173       Wyoming
174       Wyoming
175       Wyoming
Name: state, Length: 176, dtype: object

In [6]:
pd.get_dummies(dfo)

Alaska  Arizona  California  Colorado  Connecticut  Idaho  Illinois  \
0         1        0           0         0            0      0         0   
1         0        1           0         0            0      0         0   
2         0        0           1         0            0      0         0   
3         0        0           1         0            0      0         0   
4         0        0           1         0            0      0         0   
..      ...      ...         ...       ...          ...    ...       ...   
171       0        0           0         0            0      0         0   
172       0        0           0         0            0      0         0   
173       0        0           0         0            0      0         0   
174       0        0           0         0            0      0         0   
175       0        0           0         0            0      0         0   

     Indiana  Iowa  Maine  ...  Rhode Island  South Dakota  Tennessee  Utah  \
0          0     0      0  ...             0             0          0     0   
1          0     0      0  ...             0             0          0     0   
2          0     0      0  ...             0             0          0     0   
3          0     0      0  ...             0             0          0     0   
4          0     0      0  ...             0             0          0     0   
..       ...   ...    ...  ...           ...           ...        ...   ...   
171        0     0      0  ...             0             0          0     0   
172        0     0      0  ...             0             0          0     0   
173        0     0      0  ...             0             0          0     0   
174        0     0      0  ...             0             0          0     0   
175        0     0      0  ...             0             0          0     0   

     Vermont  Virginia  Washington  West Virginia  Wisconsin  Wyoming  
0          0         0           0              0          0        0  
1          0         0           0              0          0        0  
2          0         0           0              0          0        0  
3          0         0           0              0          0        0  
4          0         0           0              0          0        0  
..       ...       ...         ...            ...        ...      ...  
171        0         0           0              0          0        1  
172        0         0           0              0          0        1  
173        0         0           0              0          0        1  
174        0         0           0              0          0        1  
175        0         0           0              0          0        1  

[176 rows x 35 columns]

In [7]:
df=df.drop(['state'],axis=1)
df.head(5)

Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3         Boreal Mountain Resort         7700            500      0   
4                    Dodge Ridge         8200           1600      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  \
0        0.0          0          0     0       1       0  ...   
1        0.0          0          1     2       3       1  ...   
2        0.0          0          0     0       1       3  ...   
3        0.0          0          1     1       3       1  ...   
4        0.0          0          0     1       2       5  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              380.0      200.000000        150.000000       54.0   
4              862.0      174.873239        115.103943       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0      34.00000              152.0   
1            250.0          74.0      78.00000              104.0   
2            300.0          47.0      47.00000              107.0   
3            400.0          49.0      64.16681              150.0   
4            350.0          78.0      78.00000              140.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3           200.0         1  
4             0.0         1  

[5 rows x 25 columns]

In [8]:
df = pd.concat([df, pd.get_dummies(dfo)], axis=1)
df.head(5)

Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3         Boreal Mountain Resort         7700            500      0   
4                    Dodge Ridge         8200           1600      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          0     0       1       0  ...             0   
1        0.0          0          1     2       3       1  ...             0   
2        0.0          0          0     0       1       3  ...             0   
3        0.0          0          1     1       3       1  ...             0   
4        0.0          0          0     1       2       5  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

In [9]:
df.T

0                    1    \
Name               Hilltop Ski Area  Sunrise Park Resort   
summit_elev                    2090                11100   
vertical_drop                   294                 1800   
trams                             0                    0   
fastEight                         0                    0   
fastSixes                         0                    0   
fastQuads                         0                    1   
quad                              0                    2   
triple                            1                    3   
double                            0                    1   
surface                           2                    0   
total_chairs                      3                    7   
Runs                             13                   65   
TerrainParks                      1                    2   
LongestRun_mi                     1                  1.2   
SkiableTerrain_ac                30                  800   
Snow Making_ac                   30                   80   
daysOpenLastYear                150                  115   
yearsOpen                        36                   49   
averageSnowfall                  69                  250   
AdultWeekday                     30                   74   
AdultWeekend                     34                   78   
projectedDaysOpen               152                  104   
NightSkiing_ac                   30                   80   
clusters                          0                    1   
Alaska                            1                    0   
Arizona                           0                    1   
California                        0                    0   
Colorado                          0                    0   
Connecticut                       0                    0   
Idaho                             0                    0   
Illinois                          0                    0   
Indiana                           0                    0   
Iowa                              0                    0   
Maine                             0                    0   
Maryland                          0                    0   
Massachusetts                     0                    0   
Michigan                          0                    0   
Minnesota                         0                    0   
Missouri                          0                    0   
Montana                           0                    0   
Nevada                            0                    0   
New Hampshire                     0                    0   
New Jersey                        0                    0   
New Mexico                        0                    0   
New York                          0                    0   
North Carolina                    0                    0   
Ohio                              0                    0   
Oregon                            0                    0   
Pennsylvania                      0                    0   
Rhode Island                      0                    0   
South Dakota                      0                    0   
Tennessee                         0                    0   
Utah                              0                    0   
Vermont                           0                    0   
Virginia                          0                    0   
Washington                        0                    0   
West Virginia                     0                    0   
Wisconsin                         0                    0   
Wyoming                           0                    0   

                                             2                       3    \
Name               Yosemite Ski & Snowboard Area  Boreal Mountain Resort   
summit_elev                                 7800                    7700   
vertical_drop                                600                     500   
trams                                          0                   

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [10]:
#---- StandardScaler().fit(data).transform(data)

# It is used for the purpose of feature scaling so that all elements fall in the same range , 
# i.e , It calculates the standardised value for each element Xi of a column as follows :
#          Xi = (Xi- μ) / σ
#      where   μ=Mean of all elements of the column 
#              σ=Std Dev of all elements of the column 
# https://www.youtube.com/watch?v=PiDfEo7qpcw

In [11]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend'] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

print("X---",X,"y---",y,"scaler---",scaler,"X_scaled---",X_scaled)

X---      summit_elev  vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  \
0           2090            294      0        0.0          0          0     0   
1          11100           1800      0        0.0          0          1     2   
2           7800            600      0        0.0          0          0     0   
3           7700            500      0        0.0          0          1     1   
4           8200           1600      0        0.0          0          0     1   
..           ...            ...    ...        ...        ...        ...   ...   
171         8000            640      0        0.0          0          0     0   
172         7428            810      0        0.0          0          0     0   
173         7808           1571      0        0.0          0          0     1   
174         9663            990      0        0.0          0          0     0   
175         9500           1100      0        0.0          0          0     0   

     triple  double  s

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [12]:
# --- test_size is the size of Test data

#  Random_state can be 0 or 1 or any other integer. It should be the same value if you want to validate your processing over 
# multiple runs of the code.By the way, I have seen random_state=42 used in many official examples of scikit.The random_state parameter 
# is used for initializing the internal random number generator, which will decide the splitting of data into 
# train and test indices in your case.If random_state is None or np.random, then a randomly-initialized RandomState object is 
# returned.If random_state is an integer, then it is used to seed a new RandomState object.This is to check and validate 
# the data when running the code multiple times. Setting random_state a fixed value will guarantee that the same sequence 
# of random numbers is generated each time you run the code.

In [13]:
# --- Splitting into train , test and validation sets requires using train_test_split twice :
# X_train, X_test, y_train, y_test 
#     = train_test_split(X, y, test_size=0.2, random_state=1)

#  X_train, X_val, y_train, y_val 
#     = train_test_split(X_train, y_train, test_size=0.25, random_state=1) # 0.25 x 0.8 = 0.2

In [14]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 

X_train, X_test,y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [15]:
y

array([34.        , 78.        , 47.        , 64.16681004, 78.        ,
       75.        , 64.16681004, 59.        , 84.        , 83.        ,
       64.16681004, 79.        , 50.        , 69.        , 89.        ,
       71.        , 56.        , 84.        , 65.        , 65.        ,
       60.        , 55.        , 62.        , 42.        , 42.        ,
       47.        , 32.        , 47.        , 64.16681004, 43.        ,
       55.        , 35.        , 40.        , 45.        , 52.        ,
       39.        , 40.        , 46.        , 43.        , 55.        ,
       49.        , 32.        , 75.        , 79.        , 68.        ,
       45.        , 40.        , 64.16681004, 47.        , 35.        ,
       65.        , 49.        , 49.        , 64.16681004, 49.        ,
       34.        , 45.        , 45.        , 45.        , 35.        ,
       42.        , 45.        , 64.16681004, 45.        , 47.        ,
       44.        , 37.        , 58.        , 35.34      , 48.  

In [16]:
print("X_train---",X_train,"\nX_test---",X_test,"\ny_train---",y_train,"\ny_test---" ,y_test)

X_train--- [[-0.80077675 -0.85212798  0.         ... -0.13168538 -0.25819889
  -0.17099639]
 [-0.61050361  0.55347569  0.         ... -0.13168538 -0.25819889
  -0.17099639]
 [-0.54151026  0.19614753  0.         ... -0.13168538 -0.25819889
  -0.17099639]
 ...
 [-0.80232369 -0.9368029   0.         ... -0.13168538 -0.25819889
  -0.17099639]
 [ 1.51808048  1.26474502  0.         ... -0.13168538 -0.25819889
  -0.17099639]
 [-0.81500856 -0.64044068  0.         ... -0.13168538 -0.25819889
  -0.17099639]] 
X_test--- [[ 1.72072911  0.34178839  0.         ... -0.13168538 -0.25819889
  -0.17099639]
 [-0.7559156  -0.34407846  0.         ... -0.13168538 -0.25819889
  -0.17099639]
 [-0.64763008 -0.34407846  0.         ... -0.13168538 -0.25819889
  -0.17099639]
 ...
 [-0.76210335 -1.08921776  0.         ... -0.13168538 -0.25819889
  -0.17099639]
 [-0.49293647  0.16735806  0.         ... -0.13168538 -0.25819889
  -0.17099639]
 [-0.98950296 -1.03841281  0.         ... -0.13168538 -0.25819889
  -0.17099

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [17]:
# all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [18]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred=model.predict(X_test)
print("X_test",X_test,"\ny_pred",y_pred )


X_test [[ 1.72072911  0.34178839  0.         ... -0.13168538 -0.25819889
  -0.17099639]
 [-0.7559156  -0.34407846  0.         ... -0.13168538 -0.25819889
  -0.17099639]
 [-0.64763008 -0.34407846  0.         ... -0.13168538 -0.25819889
  -0.17099639]
 ...
 [-0.76210335 -1.08921776  0.         ... -0.13168538 -0.25819889
  -0.17099639]
 [-0.49293647  0.16735806  0.         ... -0.13168538 -0.25819889
  -0.17099639]
 [-0.98950296 -1.03841281  0.         ... -0.13168538 -0.25819889
  -0.17099639]] 
y_pred [ 5.06228294e+01  5.82395287e+01  4.70530052e+01  4.47523460e+01
  5.88567162e+01  4.75207787e+01  5.52729759e+01  8.31784477e+01
  3.69176292e+01  4.53288841e+01  7.89508324e+01  8.23692894e+01
  4.71198997e+01  6.02224388e+01  6.00442162e+01  4.25075951e+01
  4.21350365e+01  3.26814232e+01  3.43396263e+01  3.76682396e+01
  4.80862083e+01  5.08998833e+01  5.85847435e+01  5.53042259e+01
  3.27822533e+01  7.09861107e+01  6.00613060e+01  3.08232689e+01
  5.83114281e+01  6.21252708e+01  4.77

## Review Model Outcomes — Iterate over additional models as needed

In [19]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
from sklearn.metrics import explained_variance_score , mean_absolute_error
# You can plug y_test and y_pred into the functions to evaluate the model
var_score=explained_variance_score(y_test , y_pred )
mean_abs_err=mean_absolute_error(y_test , y_pred)


In [20]:
print("var_score---",var_score,"mean_abs_err---",mean_abs_err)

var_score--- -1.4876995714877353e+21 mean_abs_err--- 80222956105.49896


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [21]:
print(model.intercept_)

-20055738968.495335


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [22]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coeff=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coeff

Coefficient
summit_elev        3.961375e-01
vertical_drop      3.077587e+00
trams              3.648050e+11
fastEight          6.524533e+12
fastSixes          7.642412e+12
fastQuads          6.653224e+11
quad               7.960293e+11
triple             1.398003e+12
double             1.711529e+12
surface            1.637901e+12
total_chairs       3.117620e+12
Runs               4.519348e-01
TerrainParks       1.797363e+00
LongestRun_mi      1.089111e+00
SkiableTerrain_ac  1.676514e+00
Snow Making_ac     6.665039e-01
daysOpenLastYear   1.140747e+00
yearsOpen          3.879395e-01
averageSnowfall    1.895752e+00
AdultWeekday       9.336670e+00
projectedDaysOpen  2.112305e+00
NightSkiing_ac     9.194336e-01
clusters           1.561646e+00
Alaska             2.653125e+11
Arizona            2.653125e+11
California         8.897157e+11
Colorado           5.864375e+11
Connecticut        5.864375e+11
Idaho              6.405290e+11
Illinois           4.569013e+11
Indiana            3.741350e+11
Iowa               4.569013e+11
Maine              5.864375e+11
Maryland           2.653125e+11
Massachusetts      5.260571e+11
Michigan           9.855922e+11
Minnesota          7.352556e+11
Missouri           3.741350e+11
Montana            5.260571e+11
Nevada             2.653125e+11
New Hampshire      8.544310e+11
New Jersey         1.024099e+04
New Mexico         6.405290e+11
New York           1.120253e+12
North Carolina     4.569013e+11
Ohio               5.260571e+11
Oregon             4.569013e+11
Pennsylvania       8.544310e+11
Rhode Island       2.653125e+11
South Dakota       2.653125e+11
Tennessee          2.653125e+11
Utah               4.569013e+11
Vermont            7.352556e+11
Virginia           3.741350e+11
Washington         5.260571e+11
West Virginia      4.569013e+11
Wisconsin          8.544310e+11
Wyoming            5.864375e+11

In [23]:
coeff.sort_values(['Coefficient']).tail(10)

Coefficient
New Hampshire  8.544310e+11
California     8.897157e+11
Michigan       9.855922e+11
New York       1.120253e+12
triple         1.398003e+12
surface        1.637901e+12
double         1.711529e+12
total_chairs   3.117620e+12
fastEight      6.524533e+12
fastSixes      7.642412e+12

In [24]:
toptenCoeff=coeff.sort_values(['Coefficient'],ascending=False)
toptenCoeff.head(10)

Coefficient
fastSixes      7.642412e+12
fastEight      6.524533e+12
total_chairs   3.117620e+12
double         1.711529e+12
surface        1.637901e+12
triple         1.398003e+12
New York       1.120253e+12
Michigan       9.855922e+11
California     8.897157e+11
New Hampshire  8.544310e+11

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [25]:
df2=pd.read_csv(os.path.join(path,r'step3_output.csv') )
df2.head(5)

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3         Boreal Mountain Resort  California         7700            500   
4                    Dodge Ridge  California         8200           1600   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          1     1       3  ...   
4      0        0.0          0          0     1       2  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              380.0      200.000000        150.000000       54.0   
4              862.0      174.873239        115.103943       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0      34.00000              152.0   
1            250.0          74.0      78.00000              104.0   
2            300.0          47.0      47.00000              107.0   
3            400.0          49.0      64.16681              150.0   
4            350.0          78.0      78.00000              140.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3           200.0         1  
4             0.0         1  

[5 rows x 26 columns]

In [26]:
df2.drop(['state'],axis=1,inplace = True)
df2.head()

Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3         Boreal Mountain Resort         7700            500      0   
4                    Dodge Ridge         8200           1600      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  \
0        0.0          0          0     0       1       0  ...   
1        0.0          0          1     2       3       1  ...   
2        0.0          0          0     0       1       3  ...   
3        0.0          0          1     1       3       1  ...   
4        0.0          0          0     1       2       5  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              380.0      200.000000        150.000000       54.0   
4              862.0      174.873239        115.103943       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0      34.00000              152.0   
1            250.0          74.0      78.00000              104.0   
2            300.0          47.0      47.00000              107.0   
3            400.0          49.0      64.16681              150.0   
4            350.0          78.0      78.00000              140.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3           200.0         1  
4             0.0         1  

[5 rows x 25 columns]

In [27]:
# STANDARDIZING

X2 = df2.drop(['Name','AdultWeekend'], axis=1) 
y2 = df2['AdultWeekend'] 
scaler2 = preprocessing.StandardScaler().fit(X2)
X_scaled2=scaler2.transform(X2) 

print("X---",X2,"y---",y2,"scaler---",scaler2,"X_scaled---",X_scaled2)

X---      summit_elev  vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  \
0           2090            294      0        0.0          0          0     0   
1          11100           1800      0        0.0          0          1     2   
2           7800            600      0        0.0          0          0     0   
3           7700            500      0        0.0          0          1     1   
4           8200           1600      0        0.0          0          0     1   
..           ...            ...    ...        ...        ...        ...   ...   
171         8000            640      0        0.0          0          0     0   
172         7428            810      0        0.0          0          0     0   
173         7808           1571      0        0.0          0          0     1   
174         9663            990      0        0.0          0          0     0   
175         9500           1100      0        0.0          0          0     0   

     triple  double  s

In [28]:
# TEST TRAIN SPLIT 

y2 = y2.ravel()
X_train2, X_test2,y_train2, y_test2 = train_test_split(X_scaled2, y2, test_size=0.25, random_state=1)

In [29]:
#LINEAR REGRESSION 
model2 = lm.fit(X_train2,y_train2)
y_pred2=model2.predict(X_test2)
print("X_test",X_test2,"\ny_pred",y_pred2 )

X_test [[ 1.72072911  0.34178839  0.         ...  1.47153189 -0.94317542
   0.58111856]
 [-0.7559156  -0.34407846  0.         ... -1.77618145 -0.39083722
  -0.76462969]
 [-0.64763008 -0.34407846  0.         ... -0.10456429  0.41642631
  -0.76462969]
 ...
 [-0.76210335 -1.08921776  0.         ...  0.37557209 -0.09342434
  -0.76462969]
 [-0.49293647  0.16735806  0.         ...  0.03871718  2.77448558
  -0.76462969]
 [-0.98950296 -1.03841281  0.         ... -0.39112723  0.01279455
  -0.76462969]] 
y_pred [53.90401302 56.24971539 40.1601437  50.90139668 60.88786166 47.87237028
 52.34516943 74.90701923 37.50895989 45.64447885 75.59328597 78.04980122
 47.14896108 62.56332424 59.45968503 48.61841788 46.73699381 41.17902216
 46.12098935 40.86261933 48.44508742 54.23245195 57.71515226 61.07076437
 35.64947554 62.59536216 53.93399851 37.27614857 50.40500693 58.88561327
 53.19268748 74.74227585 95.46457734 67.71100976 56.44576039 78.39876585
 52.72642025 87.66912359 54.98441022 50.17271473 52.015

In [30]:
# VARIANCE SCORE AND ERROR
var_score2=explained_variance_score(y_test2 , y_pred2 )
mean_abs_err2=mean_absolute_error(y_test2 , y_pred2)
print("var_score2---",var_score2,"\nmean_abs_err2---",mean_abs_err2)

var_score2--- 0.5325426531233332 
mean_abs_err2--- 7.37396972614105


In [31]:
# INTERCEPT
print(model2.intercept_)

56.35202406990897


In [32]:
#COEFFICIENTS
coeff=pd.DataFrame(abs(lm.coef_), X2.columns, columns=['Coefficient'])
coeff.sort_values(['Coefficient'],ascending=False)
coeff.head(10)

Coefficient
summit_elev    2.988854e+00
vertical_drop  2.687827e+00
trams          0.000000e+00
fastEight      1.776357e-15
fastSixes      8.881784e-16
fastQuads      6.012949e-01
quad           2.053486e+00
triple         5.133334e-01
double         5.179764e-01
surface        1.312435e+00

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [33]:
df3=pd.read_csv(os.path.join(path,r'step3_output.csv') )
df3.head(5)

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3         Boreal Mountain Resort  California         7700            500   
4                    Dodge Ridge  California         8200           1600   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          1     1       3  ...   
4      0        0.0          0          0     1       2  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              380.0      200.000000        150.000000       54.0   
4              862.0      174.873239        115.103943       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0      34.00000              152.0   
1            250.0          74.0      78.00000              104.0   
2            300.0          47.0      47.00000              107.0   
3            400.0          49.0      64.16681              150.0   
4            350.0          78.0      78.00000              140.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3           200.0         1  
4             0.0         1  

[5 rows x 26 columns]

In [35]:
df3.drop(['state','summit_elev'],axis=1,inplace=True)
df3.columns

Index(['Name', 'vertical_drop', 'trams', 'fastEight', 'fastSixes', 'fastQuads',
       'quad', 'triple', 'double', 'surface', 'total_chairs', 'Runs',
       'TerrainParks', 'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac', 'clusters'],
      dtype='object')

In [36]:
#STANDARDIZING
X3 = df3.drop(['Name','AdultWeekend'], axis=1) 
y3 = df3['AdultWeekend'] 
scaler3 = preprocessing.StandardScaler().fit(X3)
X_scaled3=scaler3.transform(X3) 

# print("X---",X3,"y---",y3,"scaler---",scaler3,"X_scaled---",X_scaled3)

In [37]:
# TEST TRAIN SPLIT 

y3 = y3.ravel()
X_train3, X_test3,y_train3, y_test3 = train_test_split(X_scaled3, y3, test_size=0.25, random_state=1)

In [38]:
#LINEAR REGRESSION 
model3 = lm.fit(X_train3,y_train3)
y_pred3=model3.predict(X_test3)
# print("X_test",X_test3,"\ny_pred",y_pred3 )

In [39]:
# VARIANCE SCORE AND ERROR
var_score3=explained_variance_score(y_test3 , y_pred3 )
mean_abs_err3=mean_absolute_error(y_test3 , y_pred3)
print("var_score3---",var_score3,"\nmean_abs_err3---",mean_abs_err3)

var_score3--- 0.5266658380003132 
mean_abs_err3--- 7.645321928693992


In [40]:
#INTERCEPT
print(model3.intercept_)

56.20523908248818


In [41]:
#COEFFICIENTS
coeff=pd.DataFrame(abs(lm.coef_), X3.columns, columns=['Coefficient'])
coeff.sort_values(['Coefficient'],ascending=False)
coeff.head(10)

Coefficient
vertical_drop  2.652931e+00
trams          3.108624e-15
fastEight      2.886580e-15
fastSixes      2.664535e-15
fastQuads      7.917712e-01
quad           2.003488e+00
triple         3.262767e-01
double         6.143749e-01
surface        1.026970e+00
total_chairs   6.245785e-01

In [42]:
print("var_score---",var_score,"mean_abs_err---",mean_abs_err)
print("\nvar_score2---",var_score2,"mean_abs_err2---",mean_abs_err2)
print("\nvar_score3---",var_score3,"mean_abs_err3---",mean_abs_err3)

var_score--- -1.4876995714877353e+21 mean_abs_err--- 80222956105.49896

var_score2--- 0.5325426531233332 mean_abs_err2--- 7.37396972614105

var_score3--- 0.5266658380003132 mean_abs_err3--- 7.645321928693992


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -1.4876995714877353e+21 |80222956105.49896  |-|
| Model 2. | 0.5325426531233332|7.37396972614105 |'state'|
| Model 3. |0.5266658380003132 |7.645321928693992 |'state','summit_elev','base_elev'|

Model Selection:

Model 2 is the best model since its 
-MAE is minimum in comparison to Model 1 and Model 3 and
-Variance score is near 1.0 i.e. greater compared to other models